jupyterhub/spawner.py

```diff
@@ -994,6 +994,9 @@ def get_env(self):
             # this should only occur in mock/testing scenarios
             base_url = '/'
 
+        if self.port < 1 or self.port > 65535:
```

Comment:
This is a breaking change that takes away the option of letting the launched process pick the port, as is done in e.g. batchspawner, which phones home the resulting port after launch - a pattern I think we want _more_ spawners to be able to do, since it will select ports more reliably than a random number on the Hub's system which is typically not the same port namespace.

Ideally, this could be communicated better and fail _when not supported_, which is most of the time.

How about this:

- validate `0 <= port <= 65535`
- if port is 0, _log_ that the launched process will be picking the port, and that the Spawner better be able to handle that.

Or, we could have a small breaking change that defines an overrideable method `validate_port` that can be overridden to allow for port 0 (the default would be your check). Maybe even with a boolean attribute on the spawner class to indicate whether port 0 should be allowed.

WDYT?

Reply:
The reason that I raise an error instead of logging an error is that in the case of `port == 0` an invalid `JUPYTERHUB_SERVICE_URL` will be generated on the next lines.  Invalid in the sense that `jupyter/docker-stacks` is tripping over it and starting the server on port 80 instead of the default port 8888 or even an explicitely set JUPYTER_PORT.

Maybe part of the solution is  to add a check in `nbviewer` to make sure that a valid port is used? I can create a pull request for this commit: https://github.com/twalcari/nbviewer/commit/ef2ac83d7fe8b006f53ec6e0c0eb7b23fdc8b44e

Reply:
It doesn't produce an invalid string, though. It produces a string instructing the server to bind on port 0, which is asking the server to pick its own port. To work properly with JupyterHub, this must be _combined_ with the logic required to pass the port info back to the Hub after starting, as is done in BatchSpawner.

For _services_, which don't allow custom spawners, the `0->random port` above will fix this. The problem with the validation here that it affects _all_ Spawners, prohibiting a valid use case, when it should only be affecting the service case, and thus more appropriately confined to the Service class.

> Maybe part of the solution is to add a check in nbviewer to make sure that a valid port is used? 

Yeah, I think that's a good idea.